core/src/main/kotlin/core/ems/service/exercise/dir/ReadDir.kt

```diff
@@ -224,18 +223,39 @@ class ReadDirController {
                         dir.id.toString(),
                         ex.title,
                         dir.access,
+                        // If current dir is shared, then children is also shared. Root dir is not shared.
+                        if (currentDir?.isShared == true) true else isDirectoryShared(dir.id, caller),
                         ex.graderType,
                         if (ex.usedOnCourse) courseCount else 0,
                         ex.createdAt,
-                        ex.modifiedAt
+                        ex.createdBy,
+                        ex.modifiedAt,
+                        ex.modifiedBy
                     )
                 }
 
             Resp(currentDir, childDirs, childExercises)
         }
     }
 
-    private fun selectThisDir(dirId: Long?, currentDirAccess: DirAccessLevel?): DirResp? {
+    private tailrec fun isDirectoryShared(dirId: Long, caller: EasyUser): Boolean {
+        val callerGroupId = getImplicitGroupFromAccount(caller.id)
+        val accessCountNotZero = GroupDirAccess
+            .select { GroupDirAccess.dir eq dirId and (GroupDirAccess.group neq callerGroupId) }
+            .count() > 0
+
+        val anyAccess = Dir.select { Dir.id eq dirId }.firstNotNullOfOrNull { it[Dir.anyAccess] } != null
+        val parent = Dir.select { Dir.id eq dirId }.firstNotNullOfOrNull { it[Dir.parentDir]?.value }
+
+
+        return when {
+            parent == null -> accessCountNotZero || anyAccess
+            anyAccess || accessCountNotZero -> true
+            else -> isDirectoryShared(parent, caller)
```

Comment:
If we're looking at direct children then there's no need to look at parents since we already did when looking at currentDir.